lib/proxy.js

```diff
@@ -370,8 +370,14 @@ Proxy.prototype._onHttpServerConnectData = function(req, socket, head) {
 
   function makeConnection(port) {
     // open a TCP connection to the remote host
-    var conn = net.connect(port, function() {
+    var conn = net.connect({
+      port: port,
+      allowHalfOpen: true
```

Comment:
Yes @freewil , allowHalfOpen combined with the conn.on('finish', ...) handler just below it handles another type of socket error. When I remove that code and run the tests (OS X, node 11.9) they occasionally fail with this:
```
PROXY_TO_PROXY_SOCKET_ERROR on : { Error: This socket has been ended by the other party
    at Socket.writeAfterFIN [as write] (net.js:400:12)
    at Socket.ondata (_stream_readable.js:693:20)
    at Socket.emit (events.js:197:13)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at Socket.Readable.push (_stream_readable.js:224:10)
    at TCP.onStreamRead [as onread] (internal/stream_base_commons.js:145:17) code: 'EPIPE' }
````
From my understanding this is a variation of the ECONNRESET error. Doesn't occur on every run, it all comes down to timing. 
An alternative approach would have been to filter out also EPIPE errors in the filtering function I added - but I found this to be cleaner since it takes care to close the client socket.

Inspired by [this post](https://stackoverflow.com/questions/51792579/how-to-prevent-the-econnreset-when-tunnelling-https-request-and-the-remote-serve/51866761) on stackoverflow.